### PR TITLE
Octahedral Normal/Tangent Compression

### PR DESCRIPTION
**Pull request by The-O-King**
*Bugsquad edit: `master` version of https://github.com/godotengine/godot/pull/46800.*

Implement an import option to enable Octahedral Normals in Godot 4.0

### CODE REVIEW

File: servers/rendering_server.h
Comment (The-O-King) [inline servers/rendering_server.h:75]: I actually think that there's a new function in Vector3 which also does octahedral encoding - it might be good to consolidate those implementations (although I think that implementation maps it to UNORM rather than SNORM as we do here but that's not difficult to figure out, would just take some additional changes elsewhere to support UNORM values)

File: servers/rendering_server.cpp
Comment (The-O-King) [inline servers/rendering_server.cpp:322]: Not sure if this should live here; the SIGN implementation changed from 3.x to 4.0 where now SIGN returns 0 when 0 is passed in (which matches what glsl does but is not what we want here haha)

File: scene/resources/mesh.cpp
Comment (The-O-King) [inline scene/resources/mesh.cpp:827]: In 3.x we had the option to use octahedral encoding with 16 bit values and 8 bit values for additional memory bandwidth savings - is this something we will want to support here as well (either now or in the future?)

File: scene/resources/mesh.cpp
Comment (clayjohn) [inline scene/resources/mesh.cpp:830]: Should this be an ``int8_t``? it looks like it could be a copy-paste error

Comment (The-O-King) [reply scene/resources/mesh.cpp]: Thanks for catching this!

File: servers/rendering_server.h
Comment (The-O-King) [reply servers/rendering_server.h]: I went ahead and implemented things using the existing octahedron implementation for both the normals and tangents! I added the tangent-specific functions to Vector3 as well

As for decoding Snorm vs Unorm, the only thing that really changes is that we have to do two additional Multiply/Add operations in the shaders to decode the normals back to cartesian, not sure how sensitive we are to the performance implications, but wanted to note it at least

let me know what you all think!

**Comment by The-O-King** [comment]
Wanted to bump this to see if this was something we were looking to get into the beta? I saw a message about the feature freeze and I know this change might be a little wide reaching so just let me know if we need to hold off on it for another release?

Thanks everyone :)

**Comment by Calinou** [comment]
> Wanted to bump this to see if this was something we were looking to get into the beta? I saw a message about the feature freeze and I know this change might be a little wide reaching so just let me know if we need to hold off on it for another release?

I think the goal is still to set octahedral compression into 4.0, as it's a feature that is present in 3.x and we should avoid functional regressions when possible.

**Review by clayjohn — CHANGES_REQUESTED** [review]
Looks good to me! 

Before merging we will also need to add the corresponding changes to the GLES3 renderer as well.

**Comment by reduz** [comment]
We really should get this merged before Beta.

**Comment by The-O-King** [comment]
Thanks for the feedback! I'm at SIGGRAPH this week actually - but I should be able to also implement this in GLES3 pretty quickly by the end of the upcoming weekend if that works

**Comment by The-O-King** [comment]
Hi all! Just got a chance to get this into GLES3 - it was a little tricky to validate because I think there's something wrong with the normal map implementation on GLES3 right now? I checked in RenderDoc and the values I expect to come out of the scene shader are correct (and match what I saw in my vulkan implementation which renders correctly) so wanted to see if this was a known issue (or perhaps it has yet to be implemented?) It seems this behavior also occur on a clean master branch as well

Without normal mapping everything renders correctly though on GLES3!

**Comment by clayjohn** [comment]
> Hi all! Just got a chance to get this into GLES3 - it was a little tricky to validate because I think there's something wrong with the normal map implementation on GLES3 right now? I checked in RenderDoc and the values I expect to come out of the scene shader are correct (and match what I saw in my vulkan implementation which renders correctly) so wanted to see if this was a known issue (or perhaps it has yet to be implemented?) It seems this behavior also occur on a clean master branch as well
> 
> Without normal mapping everything renders correctly though on GLES3!

I haven't tested normal mapping extensively in GLES3 yet, so my guess is there is an existing issue in GLES3 rather than the problem being with your code.

**Review by clayjohn — APPROVED** [review]
Looks great to me! I'd like @reduz to approve the changes to vector3 before merging.

File: scene/resources/mesh.cpp
Comment (reduz) [reply scene/resources/mesh.cpp:933]: Maybe at some point, but in general this adds a lot of complexity and not a lot of win because on the devices where you save the most bandwidth on mobile you rarely use normalmapping anyway, and this requires different shader versions to decode that, which are far more expensive on Vulkan (pipelines..) so for now probably better not.

**Review by reduz — APPROVED** [review]

**Review by reduz — APPROVED** [review]

**Comment by reduz** [comment]
I think there is one thing missing, which is the softbody code (soft_dynamic_body.cpp:85), which does encoding on CPU.

**Comment by clayjohn** [comment]
> I think there is one thing missing, which is the softbody code (soft_dynamic_body.cpp:85), which does encoding on CPU.

For reference, change will be here: https://github.com/godotengine/godot/blob/b35ff86c176fef561557f2dd96b597c9c27c4e5e/scene/3d/soft_dynamic_body_3d.cpp#L87-L95

And also in Sprite 3D:
https://github.com/godotengine/godot/blob/b35ff86c176fef561557f2dd96b597c9c27c4e5e/scene/3d/sprite_3d.cpp#L561-L583
https://github.com/godotengine/godot/blob/b35ff86c176fef561557f2dd96b597c9c27c4e5e/scene/3d/sprite_3d.cpp#L928-L950

**Comment by The-O-King** [comment]
Finished up getting those supported! Do we also have blendshapes to support too? I took a quick look but doesn't seem like they've been implemented yet but wanted to make sure I wasn't missing anything

File: scene/3d/soft_dynamic_body_3d.cpp
Comment (The-O-King) [inline scene/3d/soft_dynamic_body_3d.cpp:93]: I know the code surrounding these changes uses these numbers directly; but would it be preferred to change these "magic numbers" into enums (for example stdint.h has enums for UINT16_MAX) for readability or should we keep this implementation as is?

Comment (fire) [reply scene/3d/soft_dynamic_body_3d.cpp:93]: I guess you can do constexp and pow 2 to generate these.

Comment (clayjohn) [reply scene/3d/soft_dynamic_body_3d.cpp:93]: I would leave as-is. It is quite obvious from context what you are doing.

**Comment by clayjohn** [comment]
> Finished up getting those supported! Do we also have blendshapes to support too? I took a quick look but doesn't seem like they've been implemented yet but wanted to make sure I wasn't missing anything

Good thinking. It looks like we will need to add it here as well:

https://github.com/godotengine/godot/blob/0c5f254956f0115e363ce08045dd178dc30b54f8/servers/rendering/renderer_rd/shaders/skeleton.glsl#L134

**Comment by The-O-King** [comment]
> > Finished up getting those supported! Do we also have blendshapes to support too? I took a quick look but doesn't seem like they've been implemented yet but wanted to make sure I wasn't missing anything
> 
> Good thinking. It looks like we will need to add it here as well:
> 
> https://github.com/godotengine/godot/blob/0c5f254956f0115e363ce08045dd178dc30b54f8/servers/rendering/renderer_rd/shaders/skeleton.glsl#L134

Ah perfect it was in the glsl files and I just missed it! I went and added the necessary encode/decode functions! It's a little bigger than what was there previously due to the difference with the tangent's encoding of the binormal sign - if it helps with readability I can reduce the number of functions that are there by inlining the additional work for the tangents for example?  - let me know what you think of the implementation :)

**Comment by akien-mga** [comment]
Thanks a ton!

And congrats for finally getting rid of your "First-time contributor" badge... seems like GitHub didn't count all your great `3.x` contributions as valid :(
image

**Comment by The-O-King** [comment]
Oh wow it's not often that you get a chance to make two first contributions hehe :P I'm glad I could do my small part in both branches 😄 

Thanks all for the help getting this landed! I'm excited to see how we can iterate on this and what other cool new features we can do next :)

**Comment by Xyotic** [comment]
Does this affect how we need to setup normals/tangets for procedural ArrayMeshes?

**Comment by clayjohn** [comment]
@Xyotic No. But it will affect how you update normals/tangents if you call [RenderingServer.mesh_surface_update_vertex_region](https://docs.godotengine.org/en/latest/classes/class_renderingserver.html#class-renderingserver-method-mesh-surface-update-vertex-region)

**Comment by Xyotic** [comment]
Alright thanks.
This should probably be documented tho.
There isnt really a way to tell how to use this method.

File: scene/resources/mesh.cpp
Comment (nikitalita) [inline scene/resources/mesh.cpp:941]: You changed the offset value here; we're getting 2 bytes from `src` here (src is `int8_t`, and we're getting `src[0]` and `src[1]`), so the source offset needs to be incremented by 2 bytes. You're incrementing by 4 bytes here.

File: scene/resources/mesh.cpp
Comment (nikitalita) [inline scene/resources/mesh.cpp:974]: you changed it here, too. `sizeof(float) = 4`, src is `float`, we're getting `src[0]`, `src[1]`, and `src[2]`. 12 bytes total. You're only incrementing by 4 bytes.

File: scene/resources/mesh.cpp
Comment (nikitalita) [inline scene/resources/mesh.cpp:1022]: same issue as above, except that it's 16 bytes total, and you changed it to 4.

File: scene/resources/mesh.cpp
Comment (nikitalita) [inline scene/resources/mesh.cpp:989]: same issue as above

**Pull request merged**
